File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/18f4c797aaf57b3947a6000d6b07af4c.py
N = 1000009
M = 10009
Z = 26

A = input().strip()
Q = []
have = set()
ans = {}

n = len(A)
m = int(input())

for _ in range(m):
    query = input().strip()
    q = 0
    for c in query:
        q |= (1 << (ord(c) - ord('a')))
    Q.append(q)
    have.add(q)

stk = [0] * Z
nn = 0
vis = [False] * Z

def push(x):
    global nn
    if not vis[x]:
        vis[x] = True
        stk[nn] = x
        nn += 1
    else:
        j = 0
        while stk[j] != x:
            j += 1
        for k in range(j, nn - 1):
            stk[k] = stk[k + 1]
        stk[nn - 1] = x

for i in range(n):
    push(A[i])
    s = 0
    for j in range(nn, -1, -1):
        s |= (1 << stk[j])
        if s & (1 << A[i + 1]):
            break
        if s in have:
            if s in ans:
                ans[s] += 1
            else:
                ans[s] = 1

for i in range(m):
    print(ans[Q[i]])

File: output/GPT4o-Mini/xCodeEval_4000/C++/C/18f4c797aaf57b3947a6000d6b07af4c.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <set>

#define N 1000009
#define M 10009
#define Z 26

char A[N], buf[Z];
int Q[M];
std::set<int> have;
std::map<int, int> ans;
int n, m;

int stk[Z], nn;
bool vis[Z];

void push(int x) {
    if (!vis[x]) {
        vis[stk[nn++] = x] = true;
    } else {
        int j;
        for (j = 0;; j++) if (stk[j] == x) break;
        for (; j < nn - 1; j++) stk[j] = stk[j + 1];
        stk[nn - 1] = x;
    }
}

int main() {
    #ifndef ONLINE_JUDGE
    freopen("in.txt", "r", stdin);
    #endif

    n = strlen(fgets(A, N, stdin));
    for (int i = 0; i < n; i++) A[i] -= 'a';
    A[n] = Z;

    for (int i = 0; i < m; i++) {
        fgets(buf, Z, stdin);
        int cur = 0;
        for (char *p = buf; *p; p++) {
            Q[i] |= (1 << (*p - 'a'));
        }
        have.insert(Q[i]);
    }

    for (int i = 0; i < n; i++) {
        push(A[i]);
        int s = 0;
        for (int j = nn - 1; j >= 0; j--) {
            s |= (1 << stk[j]);
            if (s & (1 << A[i + 1])) break;
            if (have.count(s)) ans[s]++;
        }
    }

    for (int i = 0; i < m; i++) {
        printf("%d\n", ans[Q[i]]);
    }
}
